test/mock-agent.js

```diff
@@ -0,0 +1,1734 @@
+
+    const response = await getResponse(body)
+    t.strictEqual(response, 'hello')
+  } catch (err) {
+    t.fail(err.message)
+  }
+})
+
+test('MockAgent - enableNetConnect with a function should allow all original dispatches to be called if mockDispatch not found', async (t) => {
+  t.plan(5)
+
+  const server = createServer((req, res) => {
+    t.strictEqual(req.url, '/foo')
+    t.strictEqual(req.method, 'GET')
+    res.setHeader('content-type', 'text/plain')
+    res.end('hello')
+  })
+  t.tearDown(server.close.bind(server))
+
+  await promisify(server.listen.bind(server))(0)
+
+  const baseUrl = `http://localhost:${server.address().port}`
+
+  const mockAgent = new MockAgent()
+  setGlobalAgent(mockAgent)
+  t.tearDown(mockAgent.close.bind(mockAgent))
+
+  const mockPool = mockAgent.get(baseUrl)
+  mockPool.intercept({
+    path: '/wrong',
+    method: 'GET'
+  }).reply(200, 'foo')
+
+  try {
+    mockAgent.enableNetConnect((value) => value === `localhost:${server.address().port}`)
+
+    const { statusCode, headers, body } = await request(`${baseUrl}/foo`, {
+      method: 'GET'
+    })
+    t.strictEqual(statusCode, 200)
+    t.strictEqual(headers['content-type'], 'text/plain')
+
+    const response = await getResponse(body)
+    t.strictEqual(response, 'hello')
+  } catch (err) {
+    t.fail(err.message)
+  }
+})
+
+test('MockAgent - enableNetConnect with an unknown input should throw', async (t) => {
+  t.plan(1)
+
+  const mockAgent = new MockAgent()
+  setGlobalAgent(mockAgent)
+  t.tearDown(mockAgent.close.bind(mockAgent))
+
+  const mockPool = mockAgent.get('http://localhost:9999')
+  mockPool.intercept({
+    path: '/wrong',
+    method: 'GET'
+  }).reply(200, 'foo')
+
+  try {
+    mockAgent.enableNetConnect({})
+
+    t.fail('should not complete')
+  } catch (err) {
+    t.strictEqual(err.message, 'Unsupported matcher. Must be one of String|Function|RegExp.')
+  }
+})
+
+test('MockAgent - disableNetConnect should throw if dispatch not found by net connect', async (t) => {
+  t.plan(1)
+
+  const server = createServer((req, res) => {
+    t.strictEqual(req.url, '/foo')
+    t.strictEqual(req.method, 'GET')
+    res.setHeader('content-type', 'text/plain')
+    res.end('hello')
+  })
+  t.tearDown(server.close.bind(server))
+
+  await promisify(server.listen.bind(server))(0)
+
+  const baseUrl = `http://localhost:${server.address().port}`
+
+  const mockAgent = new MockAgent()
+  setGlobalAgent(mockAgent)
+  t.tearDown(mockAgent.close.bind(mockAgent))
+
+  const mockPool = mockAgent.get(baseUrl)
+  mockPool.intercept({
+    path: '/wrong',
+    method: 'GET'
+  }).reply(200, 'foo')
+
+  try {
+    mockAgent.disableNetConnect()
+
+    await request(`${baseUrl}/foo`, {
+      method: 'GET'
+    })
+    t.fail('should not complete the request')
+  } catch (err) {
+    t.strictEqual(err.message, `Unable to find mock dispatch and real dispatches are disabled for http://localhost:${server.address().port}`)
+  }
+})
```

Comment:
Can you please add a few tests that use `MockAgent` directly?

Reply:
Sure, I'll make sure this is also true for MockPool and MockClient where appropriate!